### ISSUE TITLE
How do I get GraphQL Date mapped to my new DateTimeOffsetGraphType?

### ISSUE THREAD

**Issue by SapphireBrand** [graphql-dotnet/graphql-dotnet#687]
I copied the DateTimeOffsetGraphType from #665, and then declared my schema object as follows:

    internal class OrdersServiceSchema : Schema
    {
        public OrdersServiceSchema()
        {
            RegisterType<DateTimeOffsetGraphType>();
            GraphTypeRegistry.Register(typeof(DateTimeOffset), typeof(DateTimeOffsetGraphType));
            GraphTypeRegistry.Register(typeof(DateTime), typeof(DateTimeOffsetGraphType));

            Query = new OrdersServiceQuery();
        }
    }

I am hitting the Register calls in the debugger.

My query has an argument that is of GraphQL type Date!, and in OrdersServiceQuery I have the following field definition:
       
```
FieldDelegate<NonNullGraphType<ListGraphType<NonNullGraphType<ShipmentObjectGraphType>>>>(
    "shipments",
    arguments: new QueryArguments(
        new QueryArgument<DateTimeOffsetGraphType> { Name = "from", Description = "optional start of date/time range" },
        new QueryArgument<DateTimeOffsetGraphType> { Name = "to", Description = "optional end of date/time range" }
    ),
    resolve: new Func<ResolveFieldContext, DateTimeOffset?, DateTimeOffset?, Task<IEnumerable<IShipment>>>((context, from, to)
        => QueryWrapper(() => Task.FromResult(ShipmentFactory.Instance.Shipments(new ShipmentsQueryGenerator{DateStart = from, DateEnd = to }))))
);
```

When the server executes the query, I see two constructor calls for DateTimeOffsetGraphType made from DefaultDependencyResolver, which I believe must be for the two arguments. Then I expected calls to DateTimeOffsetGraphType member functions to convert the Date objects.

What actually happens: I get an error message from GraphQL.Validation.Rules.VariablesInAllowedPosition: "Variable "$from" of type "Date!" used in position expecting type "DateTimeOffset"."

So I am missing something. How is the Date object tied to something that can be converted into a GraphType?

The code from DateTimeOffsetGraphType is attached: [DateTimeOffsetGraphType.cs.txt](https://github.com/graphql-dotnet/graphql-dotnet/files/2080127/DateTimeOffsetGraphType.cs.txt)

**Comment by joemcbride** [comment]
> My query has an argument that is of GraphQL type Date!

You should be using `DateTimeOffset` vs. `Date!`, this is what the error is trying to tell you.

> GraphQL.Validation.Rules.VariablesInAllowedPosition: "Variable "$from" of type "Date!" used in position expecting type "DateTimeOffset"."

You're defining the arguments to be of `DateTimeOffsetGraphType`.  So that is the type you need to use in your query.

```
arguments: new QueryArguments(
  new QueryArgument<DateTimeOffsetGraphType> { Name = "from", Description = "optional start of date/time range" },
  new QueryArgument<DateTimeOffsetGraphType> { Name = "to", Description = "optional end of date/time range" }
),
```

```
query Shipments($from: DateTimeOffset, $to: DateTimeOffset) {
  shipments(to: $to, from: $from) {
    id
  }
}
```

**Comment by SapphireBrand** [comment]
Oh, in the query. I can declare my arguments to be things that are not built into GraphQL. I didn't know that, but it's a useful feature. I will try that.

**Comment by joemcbride** [comment]
Did that get resolved with #693 ?

**Comment by SapphireBrand** [comment]
I confess that every question I have asked over the last few days has been because of problems in my own code. :-) I worked them out by cloning master and stepping through in the debugger. Understanding the problems in that way was both easy and enlightening, and I highly recommend the experience.

I opened #693 thinking that this exception was related to how variables were converted into value types, but I was mistaken about that.

In the debugger, I found that this exception occurred because I had a DateTime field that was being pushed into the DateTimeOffsetObjectGraph value chain because I didn't declare something properly. I thought I had changed all of the DateTime fields to DateTimeOffset, but I missed one. The exception here went away after I declared my field as a DateTimeOffset.

There was actually a problem (also in my code) about how DateTimeOffset variables are converted into value types. So that is how #693 came to be opened and then closed.

So I am closing this issue.
